### PR TITLE
Adding B3 propagator

### PR DESCRIPTION
**Pull request by TomRoSystems**
Basing on https://github.com/open-telemetry/opentelemetry-cpp/blob/880e392e10b6d8e964ade2e22d7231950d6d2443/api/include/opentelemetry/trace/propagation/http_trace_context.h

### CODE REVIEW

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (lalitb) [inline api/include/opentelemetry/trace/propagation/b3_propagator.h:205]: Probably header description can be like :
`B3Propogator` class provides interface that enables extracting and injecting context into single header of HTTP Request.
`B3PropogatorMultiHeader` class provides interface that enables extracting and injecting context into multiple headers of HTTP Request ?

**Review by lalitb — COMMENTED** [review]
Thanks for working on this. Looks good to me in general. Would wait for other's feedback for proper place for it:

`repo_root/api/include/opentelemetry/trace/propagation/` or
`repo_root/propagators`

As per Opentelemetry specs (https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/api-propagators.md#propagators-distribution ) - only `W3C Tracecontext` and `W3C Baggage` MAY be distributed as part of API package, and rest as extension packages.

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (lalitb) [inline api/include/opentelemetry/trace/propagation/b3_propagator.h:81]: As B3 trace_id could also be 16 hex chars, and we are always trying to convert 32 hex chars. Should there be any check to avoid that, or else put some comment that it can fail.

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (lalitb) [inline api/include/opentelemetry/trace/propagation/b3_propagator.h:24]: nit: we follow google C++ style guide (https://google.github.io/styleguide/cppguide.html#Constant_Names) which recommends `const` to be prefixed with `k`.

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (lalitb) [inline api/include/opentelemetry/trace/propagation/b3_propagator.h:211]: nit: though no strong preference, just wondering if `B3PropagatorSingleHeader` would be more readable as class name ?

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (bogdandrutu) [reply api/include/opentelemetry/trace/propagation/b3_propagator.h:77]: >Either have trace::SpanId and trace::TraceId constructor take base16 string as argument ( we already have method to return base16 trace::SpanId::ToLowerBase16 ), or move this functionality to some common place ( HTTPTextFormat ? ) as separate PR.

I think this is the right thing to do to avoid duplicate code.

Comment (TomRoSystems) [reply api/include/opentelemetry/trace/propagation/b3_propagator.h:77]: I agree that having this in constructor would be handy. But what if hex value is malformed? Should then entire object be "silently" set as invalid or throw an exception?
Or should I create corresponding method like "FromLowerBase16" which takes as an argument string_view?
I will create separate PR to remove duplication once this gets merged.

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (TomRoSystems) [reply api/include/opentelemetry/trace/propagation/b3_propagator.h:81]: Yes! Thanks for pointing this out. I'll create fix and test for such case.

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (TomRoSystems) [reply api/include/opentelemetry/trace/propagation/b3_propagator.h:211]: Just worried that this might be misleading as at the moment this class inherits from B3PropagatorExtractor which extracts both [single and multiline headers](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/api-propagators.md#b3-extract) and injects as in specification:  [MUST default to injecting B3 using the single-header format](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/api-propagators.md#b3-inject) so I considered this class as "main" one as it sticks to default behaviour and hence it's just B3Propagator. Whereas class B3PropagatorSingleHeader differs only in a way how it is extracting headers (with same injecting strategy) and it's not default one so that's why it has longer name. But if there will be more comments for changing name I'm happy to change.

**Comment by TomRoSystems** [comment]
> Thanks for working on this. Looks good to me in general. Would wait for other's feedback for proper place for it:
> 
> `repo_root/api/include/opentelemetry/trace/propagation/` or
> `repo_root/propagators`
> 
> As per Opentelemetry specs (https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/api-propagators.md#propagators-distribution ) - only `W3C Tracecontext` and `W3C Baggage` MAY be distributed as part of API package, and rest as extension packages.

Thank you for review.

Any more opinions on this one? In go there's [propagation](https://github.com/open-telemetry/opentelemetry-go/tree/master/propagation) directory same in [dotnet](https://github.com/open-telemetry/opentelemetry-dotnet/tree/master/src/OpenTelemetry.Api/Context/Propagation) whereas Java has [extensions directory on top level](https://github.com/open-telemetry/opentelemetry-java/tree/master/extensions/trace-propagators/src/main/java/io/opentelemetry/extension/trace/propagation).

**Review by lalitb — APPROVED** [review]

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (seemk) [inline api/include/opentelemetry/trace/propagation/b3_propagator.h:135]: Perhaps we should move hex to int conversion to a shared part? Currently the method (+ comment) is duplicated by b3 and http propagators.

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (seemk) [inline api/include/opentelemetry/trace/propagation/b3_propagator.h:169]: This will allow an invalid length trace id and span id to pass through, e.g. `a-b` will produce `000000000000000a` as the trace ID.

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (seemk) [inline api/include/opentelemetry/trace/propagation/b3_propagator.h:119]: This branch is never taken as `HexToInt` returns `uint8_t` (`uint8_t(-1) = int(255)`), thus a malformed hex sequence is generated.

**Review by seemk — CHANGES_REQUESTED** [review]

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (TomRoSystems) [reply api/include/opentelemetry/trace/propagation/b3_propagator.h:119]: Yep :( copied from here... https://github.com/open-telemetry/opentelemetry-cpp/blob/ce89aa10941c14a6e52e66e68e5c1831c143eaf2/api/include/opentelemetry/trace/propagation/http_trace_context.h#L120-L123

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (seemk) [inline api/include/opentelemetry/trace/propagation/b3_propagator.h:155]: Might be a nitpick, but I think we should validate the parsed trace_id, span_id lengths for validity. Say if a malformed header with too large of a trace_id gets sent, currently this logic will produce a valid `SpanContext`, but with some of the trace_id cut off - my opinion is we should just fail early in these cases.

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (TomRoSystems) [reply api/include/opentelemetry/trace/propagation/b3_propagator.h]: Fixed

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (TomRoSystems) [reply api/include/opentelemetry/trace/propagation/b3_propagator.h:155]: Thanks.
Would like to get more opinion on this topic [Robustness](https://en.wikipedia.org/wiki/Robustness_principle) vs [it's potential harmful consequences](https://tools.ietf.org/html/draft-iab-protocol-maintenance-01) from others as I have not found clear guidance on this one.

Comment (lalitb) [reply api/include/opentelemetry/trace/propagation/b3_propagator.h:155]: In case of any error in parsing `trace_id` and `span_id`, the right approach should be to set them invalid ( i.e. all zeros ) and return the SpanContext. This will ensure that SpanContext will be invalid.

File: api/include/opentelemetry/nostd/string_view.h
Comment (maxgolov) [inline api/include/opentelemetry/nostd/string_view.h:123]: Looks good. Minor concern is - since we're adding it here, it would'be been ideal to add all overloads:
```cpp
constexpr size_type find(basic_string_view v, size_type pos = 0) const noexcept;
constexpr size_type find(CharT ch, size_type pos = 0) const noexcept;
constexpr size_type find(const CharT* s, size_type pos, size_type count) const;
constexpr size_type find(const CharT* s, size_type pos = 0) const;
```

Comment (TomRoSystems) [reply api/include/opentelemetry/nostd/string_view.h:123]: I'd like to avoid [Scope creep] (https://en.wikipedia.org/wiki/Scope_creep) and complaints related to complexity for those operations.

**Comment by pyohannes** [comment]
Sorry, I'm a bit late with my comment.

> As per [Opentelemetry specs](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/api-propagators.md#propagators-distribution ) - only W3C Tracecontext and W3C Baggage MAY be distributed as part of API package, and rest as extension packages.

Would it be better to put this under `ext`, instead of `api`? This would be more true to the spec, and help to keep the API part small. [Java](https://github.com/open-telemetry/opentelemetry-java/tree/main/extensions/trace-propagators) has those in an extension directory.

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (ThomsonTan) [inline api/include/opentelemetry/trace/propagation/b3_propagator.h:46]: nit: extra space after object.

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (ThomsonTan) [inline api/include/opentelemetry/trace/propagation/b3_propagator.h:80]: `b_ptr` is unused now?

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (ThomsonTan) [inline api/include/opentelemetry/trace/propagation/b3_propagator.h:88]: Ditto.

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (ThomsonTan) [inline api/include/opentelemetry/trace/propagation/b3_propagator.h:139]: Convert return value to `int8_t` to be consistent with the declared return type?

File: api/include/opentelemetry/trace/propagation/b3_propagator.h
Comment (ThomsonTan) [inline api/include/opentelemetry/trace/propagation/b3_propagator.h:224]: Declare the length with the constants defined in the begining of this file?

**Review by ThomsonTan — APPROVED** [review]

**Comment by TomRoSystems** [comment]
@lalitb @ThomsonTan 
Guys I've created https://github.com/open-telemetry/opentelemetry-cpp/issues/539 and will move rest of work over there.

Can this be merged now?

**Comment by lalitb** [comment]
> @lalitb @ThomsonTan
> Guys I've created #539 and will move rest of work over there.
> 
> Can this be merged now?

Thanks @TomRoSystems . @pyohannes suggested on moving B3 propagator to `ext` directory. If you agree on that, we can either do it as part of #539, or have a separate issue to let someone pick it up.

**Comment by TomRoSystems** [comment]
> > @lalitb @ThomsonTan
> > Guys I've created #539 and will move rest of work over there.
> > Can this be merged now?
> 
> Thanks @TomRoSystems . @pyohannes suggested on moving B3 propagator to `ext` directory. If you agree on that, we can either do it as part of #539, or have a separate issue to let someone pick it up.

I'd like to do it as part of #539

**Pull request merged**
